server/src/main/java/io/druid/server/QueryResource.java

```diff
@@ -220,8 +222,17 @@ public void write(OutputStream outputStream) throws IOException, WebApplicationE
                 },
                 contentType
             )
-            .header("X-Druid-Query-Id", queryId)
-            .header("X-Druid-Response-Context", jsonMapper.writeValueAsString(responseContext))
+            .header("X-Druid-Query-Id", queryId);
+
+        //Limit the response-context header, see https://github.com/druid-io/druid/issues/2331
+        String responseCtxString = jsonMapper.writeValueAsString(responseContext);
```

Comment:
header bytes are encoded using ASCII, so 1 char = 1 byte
in the builder.header(String key, Object value), eventually somewhere value.toString() is called and encoded using ASCII . I tried passing a byte[] to builder.header(..) and in response I got `[B@468a1ce3`.
I have updated the comment.